api/src/main/java/jakarta/data/repository/KeysetPageable.java

```diff
@@ -0,0 +1,328 @@
+         * or <code>OrderBy</code> name pattern of the repository method.
+         * The order of results on each page follows the sort criteria
+         * and is not reversed.
+         */
+        PREVIOUS
+    }
+
+    /**
+     * Represents keyset values, which can be a starting point for
+     * requesting a next or previous page.
+     */
+    public interface Cursor {
+        /**
+         * Returns whether or not the keyset values of this instance
+         * are equal to those of the supplied instance.
+         * Cursor implementation classes must also match to be equal.
+         *
+         * @return true or false.
+         */
+        @Override
+        public boolean equals(Object o);
+
+        /**
+         * Returns the keyset value at the specified position.
+         *
+         * @param  index position (0 is first) of the keyset value to obtain.
+         * @return the keyset value at the specified position.
+         * @throws IndexOutOfBoundsException if the index is negative
+         *         or greater than or equal to the {@link #size}.
+         */
+        public Object getKeysetElement(int index);
+
+        /**
+         * Returns a hash code based on the keyset values.
+         *
+         * @return a hash code based on the keyset values.
+         */
+        @Override
+        public int hashCode();
+
+        /**
+         * Returns the number of values in the keyset.
+         *
+         * @return the number of values in the keyset.
+         */
+        public int size();
+
+        /**
+         * String representation of the keyset cursor, including the number of
+         * key values in the cursor but not the values themselves.
+         *
+         * @return String representation of the keyset cursor.
+         */
+        @Override
+        public String toString();
+    }
+
+    /**
+     * Built-in implementation of Cursor.
+     */
+    static class CursorImpl implements Cursor {
+        /**
+         * Keyset values.
+         */
+        private final Object[] keyset;
+
+        /**
+         * Constructs a keyset cursor with the specified values.
+         *
+         * @param keyset keyset values.
+         * @throws IllegalArgumentException if no keyset values are provided.
+         */
+        CursorImpl(Object... keyset) {
+            this.keyset = keyset;
+            if (keyset == null || keyset.length == 0)
+                throw new IllegalArgumentException("No keyset values were provided.");
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return this == o || o != null
+                    && o.getClass() == getClass()
+                    && Arrays.equals(keyset, ((CursorImpl) o).keyset);
+        }
+
+        public Object getKeysetElement(int index) {
+            return keyset[index];
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(keyset);
```

Comment:
Minor comment, should probably use `Arrays.hashCode` directly.

Reply:
Thanks - I had to rebase onto latest anyways due to a merge conflict with a line in repository.asciidoc, after which I spotted a couple of minor grammar errors in my changes there as well as another place to state the requirement for sort criteria to uniquely identify entities, and I also added the update to Arrays.hashCode.  This should be all of the updates that I'm aware of.